File: answers/python/exercise3/SSLClient.py
import socket
import ssl
from pathlib import Path
import pickle
import hashlib


LOCAL_HOST = 'localhost'
LOCAL_PORT = 8484
RESOURCE_DIRECTORY = Path(__file__).resolve().parent.parent / 'resources' / 'client'
CA_CERT = RESOURCE_DIRECTORY / 'ca.cert.pem'
PINNED_FILE = RESOURCE_DIRECTORY / 'pinned.hash'


def main():
    """
    Exercise 3 Client
    """
    # Create a standard TCP Socket
    sock = socket.socket(socket.AF_INET)

    # Create SSL context which holds the parameters for any sessions
    context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
    context.check_hostname = False
    context.load_verify_locations(CA_CERT)

    # We can wrap in an SSL context first, then connect
    conn = context.wrap_socket(sock, server_hostname=LOCAL_HOST)
    try:
        # Handshake - conn is an SSLSocket
        conn.connect((LOCAL_HOST, LOCAL_PORT))

        # Verify against pinned hash
        dat = conn.getpeercert(binary_form=True)
        hashalg = hashlib.sha256()
        hashalg.update(dat)
        cert_hash = hashalg.digest()

        with open(PINNED_FILE, 'rb') as pinned:
            pinned_hash = pinned.read()

        if cert_hash != pinned_hash:
            raise ssl.CertificateError("Non-pinned certificate!")

        # What parameters were established?
        print("Negotiated session using cipher suite: {0}\n".format(conn.cipher()[0]))

        # We are sending a 4 byte int, convert into bytes
        conn.send((100124796).to_bytes(4, 'big'))

        # Receive a number back from the server
        server_response = conn.recv(1024)

        # Unpickle the bytes from the server into a datatypes.BankCustomer
        customer_returned = pickle.loads(server_response)
        print(customer_returned)

    except ssl.CertificateError:
        print("The host's certificate has not been pinned by this application")
    finally:
        conn.close()


if __name__ == '__main__':
    main()

File: answers/java/exercise3/SSLClient.java
package com.tlsexercises.exercise3;

import javax.net.ssl.*;
import java.io.*;
import java.security.*;
import java.security.cert.X509Certificate;
import java.security.cert.CertificateException;
import java.util.Arrays;

/**
 * Exercise 3 Client
 */
public class SSLClient {

    private static final String REMOTE_HOST = "localhost";
    private static final int REMOTE_PORT = 8484;
    private static final String TRUSTSTORE_LOCATION = SSLClient.class.getResource("/client/ClientTrustStore.jks").getPath();
    private static final String TRUSTSTORE_PASSWORD = "experttls";

    private static final byte[] SERVER_PINNED_HASH;
    static {
        SERVER_PINNED_HASH = new byte[32];
        try {
            SSLClient.class.getResourceAsStream("/client/pinned.hash").read(SERVER_PINNED_HASH,0,32);
        } catch (Exception e) {
            throw new ExceptionInInitializerError("Unable to load server pinset.");
        }
    }

    // Entry point
    public static void main(String[] args) {
        // You can also update properties as a command line parameter e.g.
        //     -Djavax.net.ssl.keyStore="keystore_location"
        System.setProperty("javax.net.ssl.trustStore", TRUSTSTORE_LOCATION);
        System.setProperty("javax.net.ssl.trustStorePassword", TRUSTSTORE_PASSWORD);

        TrustManagerFactory tmf = null;

        try {
            // Default is likely to be PKIX, but could be SunX509 on some systems
            String def = TrustManagerFactory.getDefaultAlgorithm();
            tmf = TrustManagerFactory.getInstance(def);

            // Using null here initialises the default trust store, which in this case is loaded from the above properties
            tmf.init((KeyStore) null);
        } catch (KeyStoreException e) {
            System.err.println("Keystore Exception: " + e.getMessage());
        } catch (NoSuchAlgorithmException e) {
            System.err.println("Unable to obtain a trust manager: " + e.getMessage());
        }

        // Get hold of the default trust manager - will only return one as we only specified one algorithm above.
        X509TrustManager defaultTrustManager = (X509TrustManager) tmf.getTrustManagers()[0];

        // Wrap it in your own class.
        X509TrustManager pinningX509TrustManager = new X509TrustManager() {

            @Override
            public X509Certificate[] getAcceptedIssuers() {
                // Pass through
                return defaultTrustManager.getAcceptedIssuers();
            }

            @Override
            public void checkServerTrusted(X509Certificate[] chain,
                                           String authType) throws CertificateException {
                // Use default trust manager to verify certificate chain
                defaultTrustManager.checkServerTrusted(chain, authType);

                // The first certificate in the chain
                X509Certificate serverCert = chain[0];

                // .getEncoded returns a DER encoding of the certificate, which should match our pinned hash
                byte[] certBytes = serverCert.getEncoded();

                // Hash the certificate
                byte[] serverHash = null;
                MessageDigest digest = null;
                try {
                    digest = MessageDigest.getInstance("SHA-256");
                    serverHash = digest.digest(certBytes);
                } catch (NoSuchAlgorithmException e) {
                    throw new CertificateException("Unable to load SHA-256 digest and verify pinned certificate");
                }

                // Verify against pinned certificate hash
                if (!Arrays.equals(SERVER_PINNED_HASH, serverHash)) {
                    throw new CertificateException("The host's certificate has not been pinned by this application");
                }
            }

            @Override
            public void checkClientTrusted(X509Certificate[] chain,
                                           String authType) throws CertificateException {
                // We could validate client certs here, but it is out of scope for what we want to do
                throw new CertificateException("This trust manager does not verify client certificates");
            }
        };

        // Obtain the default socket factory
        SSLContext context = null;
        SSLSocketFactory f = null;
        try {
            context = SSLContext.getInstance("TLS");
            context.init(null, new TrustManager[] { pinningX509TrustManager }, null);
            f = context.getSocketFactory();
        } catch (Exception e) {
            System.err.println("Error establishing SSL context and factory");
            System.err.println(e.toString());
            System.exit(-1);
        }

        try {
            // Create a socket - will not connect yet
            SSLSocket socket = (SSLSocket) f.createSocket(REMOTE_HOST, REMOTE_PORT);

            // Handshake to create a session
            socket.startHandshake();

            // What parameters were established?
            System.out.println(String.format("Negotiated Session: %s", socket.getSession().getProtocol()));
            System.out.println(String.format("Cipher Suite: %s", socket.getSession().getCipherSuite()));

            // We're reading and writing objects this time
            ObjectInputStream input = new ObjectInputStream(socket.getInputStream());
            ObjectOutputStream output = new ObjectOutputStream(socket.getOutputStream());

            // Obtain bank customer from DB
            output.writeInt(100124796);
            output.flush();

            // Read the BankCustomer returned by the server.
            System.out.println("Server Response:");
            BankCustomer customer = (BankCustomer) input.readObject();

            System.out.println(customer);

            input.close();
            output.close();
            socket.close();
        } catch (Exception e) {
            System.err.println("Exception: " + e.toString());
        }
    }
}